examples/shor.py

```diff
@@ -83,9 +95,179 @@ def naive_order_finder(x: int, n: int) -> int:
     return r
 
 
-def quantum_order_finder(x: int, n: int) -> int:
-    """Computes smallest positive r such that x^r mod n == 1."""
-    raise NotImplementedError('Quantum order finder is not implemented yet.')
+class ModularExp(cirq.ArithmeticOperation):
+    """Quantum modular exponentiation.
+
+    This class represents the unitary which multiplies a given exponent of
+    x into the ancilla register. More precisely, it represents the unitary V
+    which computes modular exponentiation x^e mod n:
+
+        V|y⟩|e⟩ = |y * x^e mod n⟩ |e⟩      0 <= y < n
+        V|y⟩|e⟩ = |y⟩ |e⟩                  n <= y
+
+    where y is the ancilla register, e is the exponent register and x and n are
+    non-negative integer constants. Consequently,
+
+        V|y⟩|e⟩ = (U^e|r⟩)|e⟩
+
+    where U is the unitary defined as
+
+        U|y⟩ = |y * x mod n⟩      0 <= y < n
+        U|y⟩ = |y⟩                n <= y
+
+    in the header of this file.
+
+    Quantum order finding algorithm (which is the quantum part of the Shor's
+    algorithm) uses quantum modular exponentiation together with the Quantum
+    Phase Estimation to compute the order of x modulo n.
+    """
+
+    def __init__(self, ancilla_register: Sequence[cirq.Qid],
```

Comment:
Rename to `target`, `exponent`, `base`, `modulus`.

Reply:
Much nicer, thanks!